Source/Core/Duality/Components/Physics/RigidBody.cs

```diff
@@ -772,50 +775,65 @@ public bool PickShapes(Vector2 worldCoord, Vector2 size, List<ShapeInfo> pickedS
+				PhysicsUnit.LengthToPhysical * worldCoord,
+				PhysicsUnit.LengthToPhysical * new Vector2(worldCoord.X + size.X, worldCoord.Y),
+				PhysicsUnit.LengthToPhysical * (worldCoord + size),
+				PhysicsUnit.LengthToPhysical * new Vector2(worldCoord.X, worldCoord.Y + size.Y) }), 1);
+			Mat22 rot = new Mat22();
+			rot.SetIdentity();
+			Vector2 pos = new Vector2(0, 0);
+			FarseerPhysics.Common.Transform boxTransform = new FarseerPhysics.Common.Transform(ref pos, ref rot);
+			Manifold manifold = new Manifold();
 			int oldCount = pickedShapes.Count;
+
+			FarseerPhysics.Common.Transform bodyTransform;
+			this.body.GetTransform(out bodyTransform);
+
 			for (int i = 0; i < this.body.FixtureList.Count; i++)
 			{
 				Fixture f = this.body.FixtureList[i];
 				ShapeInfo s = f.UserData as ShapeInfo;
-
-				FarseerPhysics.Collision.AABB fAABB;
-				FarseerPhysics.Common.Transform transform;
-				this.body.GetTransform(out transform);
-				f.Shape.ComputeAABB(out fAABB, ref transform, 0);
-				
-				if (fsWorldAABB.Contains(ref fAABB))
+				PolygonShape polygonShape = f.Shape as PolygonShape;
+				if (polygonShape != null)
 				{
-					pickedShapes.Add(s);
-					continue;
+					Collision.CollidePolygons(ref manifold, boxShape, ref boxTransform, polygonShape, ref bodyTransform);
+					if (manifold.PointCount > 0)
+					{
+						pickedShapes.Add(s);
+						continue;
+					}
 				}
-				else if (!FarseerPhysics.Collision.AABB.TestOverlap(ref fsWorldAABB, ref fAABB))
-					continue;
-
-				FarseerPhysics.Collision.AABB fAABBIntersect;
-				fAABBIntersect.LowerBound = Vector2.Max(fAABB.LowerBound, fsWorldAABB.LowerBound);
-				fAABBIntersect.UpperBound = Vector2.Min(fAABB.UpperBound, fsWorldAABB.UpperBound);
 
-				Vector2 fsWorldCoordStep = PhysicsUnit.LengthToPhysical * (new Vector2(MathF.Max(s.AABB.W, 1.0f), MathF.Max(s.AABB.H, 1.0f)) * 0.05f);
-				Vector2 fsTemp = fAABBIntersect.LowerBound;
-				do
+				CircleShape circleShape = f.Shape as CircleShape;
+				if (circleShape != null)
 				{
-					if (f.TestPoint(ref fsTemp))
+					Collision.CollidePolygonAndCircle(ref manifold, boxShape, ref boxTransform, circleShape, ref bodyTransform);
+					if (manifold.PointCount > 0)
 					{
 						pickedShapes.Add(s);
-						break;
+						continue;
 					}
+				}
 
-					fsTemp.X += fsWorldCoordStep.X;
-					if (fsTemp.X > fAABBIntersect.UpperBound.X)
-					{
-						fsTemp.X = fAABBIntersect.LowerBound.X;
-						fsTemp.Y += fsWorldCoordStep.Y;
-					}
-					if (fsTemp.Y > fAABBIntersect.UpperBound.Y) break;
-				} while (true);
+				// This one is still buggy
+				//var chainShape = f.Shape as ChainShape;
+				//if (chainShape != null)
+				//{
+
+				//	for (int j = 0; j < chainShape.Vertices.Count - 1; j++)
+				//	{
+				//		var jNext = j + 1;
+				//		if (jNext >= chainShape.Vertices.Count) jNext = 0;
+				//		var edgeShape = new EdgeShape(chainShape.Vertices[j], chainShape.Vertices[jNext]);
```

Comment:
Not sure if that will fix it, but instead of creating a new edge shape every loop iteration, create it once and let the chain shape configure it. There's a `GetChildEdge` method that is public and I believe also used by Farseers internals, [here](https://github.com/AdamsLair/farseerduality/blob/65b1396bf03f83706fe740c097e56b25053d058a/Collision/Shapes/ChainShape.cs#L128).

Alternatively, if you want to keep chain shapes our of this PR for now, maybe just add a note to the commented out section regarding this, for future reference.

Reply:
I was still experimenting with this part. I already tried it with GetChildEdge but got some weird results so have to investigate that some more. Not for this PR though so Ill put a clarifying comment there